### ISSUE TITLE
Using DBContext in .NET Core 3.1 Console app with DI - Cannot insert duplicate key row in object 'dbo.Entries' with unique index 'IX_Entries_Table1Id_Table2Id'. 

### ISSUE THREAD

**Issue by djrpascu** [dotnet/efcore#20675]
I have a simple .NET Core 3.1 console app that is being used to load data into a SQL Server database. I have a method that does inserts records into a table with a unique index on 2 columns. It seems to enter records without problem until it hits the first case where the unique constraint is violated, then after that it just keeps throwing that error for each subsequent record.

I am using EF Core and trying to do things the "right" way by using Dependency Injection. I am new to DI and know that I'm doing something wrong and  hoping for some help here. I think I'm not understanding how the scope of the context works.

Main method in Program.cs - Registering DBContext and concrete class for TOIDPLoad?
```C#
           var serviceProvider = new ServiceCollection()
                .AddDbContext<TOIDPContext>(options => options.UseSqlServer(configuration.GetConnectionString("TOIDP"))
                    .EnableSensitiveDataLogging())
                .AddScoped<IGenericRepository<Entry>, GenericRepository<Entry>>()
                .AddScoped<TOIDPLoad>()
                .BuildServiceProvider();

            // CALL START PROGRAM WITH SCOPE?
            using (var scope = serviceProvider.CreateScope())
            {
                serviceProvider.GetService<TOIDPLoad>().StartProgram();
            }
```

TOIDPLoad.cs
```C#
        private TOIDPContext _TOIDPContext;

        private readonly IGenericRepository<Entry> _EntryRepository;

        public TOIDPLoad(TOIDPContext toidpContext, IGenericRepository<Entry> entryRepository)
        {
            _TOIDPContext = toidpContext;
            //_TOIDPContext.ChangeTracker.QueryTrackingBehavior = QueryTrackingBehavior.NoTracking;
        }

        public void StartProgram()
        {
            Console.WriteLine("Start TOIDP Data Move Program: {0}", DateTime.Now.ToString());

            LoadingEntries();

            //ReadEntries();

            Console.WriteLine("Finish TOIDP Data Move Program: {0}", DateTime.Now.ToString());
        }
```

LoadEntries() Method inside TOIDPLoad.cs, just separated it for readability.
```C#
public void LoadingEntries()
        {
            Console.WriteLine("Loading Entries");
            try
            {
                var entryJson = File.ReadAllText($"{_DataFolder}v_entries_detail.json");
                var v1Entries = JsonConvert.DeserializeObject<List<EntryDetailV1>>(entryJson);

                foreach (var v1Entry in v1Entries)
                {
                    var english = new Word();
                    var ilokano = new Word();
                    var definition = new Definition();
                    var userLastModified = new User();
                    var userAdded = new User();
                    try
                    {
                        english = _TOIDPContext.Words.AsNoTracking().Where(w => w.Term == v1Entry.ENG_WOR_word && w.LanguageId == 4).ToList().FirstOrDefault();
                        ilokano = _TOIDPContext.Words.AsNoTracking().Where(w => w.Term == v1Entry.ILO_WOR_word && w.LanguageId == 5).ToList().FirstOrDefault();
                        definition = _TOIDPContext.Definitions.AsNoTracking().Where(d => d.Description == v1Entry.DEF_definition).ToList().FirstOrDefault();
                        userLastModified = _TOIDPContext.Users.AsNoTracking().Where(u => u.Email == v1Entry.USE_userLastModifiedEmail).ToList().FirstOrDefault();
                        userAdded = _TOIDPContext.Users.Where(u => u.Email == v1Entry.USE_userOriginalEmail).ToList().FirstOrDefault();

                        var dateCreatedUpdated = GeneralService.ConvertOldTOIDPUnixTime(v1Entry.ENT_dateLastModified);

                        var entry = new Entry()
                        {
                            CreatedOn = dateCreatedUpdated,
                            UpdatedOn = dateCreatedUpdated,
                            WordId = ilokano.Id,
                            EnglishId = english.Id,
                            Approved = true,
                            Notes = string.IsNullOrEmpty(v1Entry.ENT_remarks) ? null : v1Entry.ENT_remarks,
                            UserLastModifiedEntryId = userLastModified.Id,
                            UserAddedEntryId = userAdded.Id
                        };

                        _TOIDPContext.Entries.Add(entry);
                        _TOIDPContext.SaveChanges();

                        //_TOIDPContext.Entry(entry).State = EntityState.Detached;

                        //_EntryRepository.Insert(entry);
                        //_EntryRepository.Commit();
                    }
                    catch (Exception exception)
                    {
                        Console.WriteLine("Error: {0} not added. \n{1}", v1Entry.ENG_WOR_word, exception.Message);
                    }
                }
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.Message);
            }
        }
```

When I hit the SaveChanges in the LoadEntries method I see records start to get inserted to the table, but it seems that when the first time the unique constraint is violated I start getting the error for every record thereafter. The duplicate key value is the same for each record, even though I'm loading different values:
```
Cannot insert duplicate key row in object 'dbo.Entries' with unique index 'IX_Entries_WordId_EnglishId'. The duplicate key value is (3cb4e4cd-4284-41ba-2d06-08d7e2ff7cc5, 9ff1d720-0315-4582-34e7-08d7e2ff7cc5).
The statement has been terminated.
```

I think it's because the DBContext isn't scoped right. Before I went on this effort to use DI, I had the Add and SaveChanges calls inside of a using statement and that worked ok. But I read that with DI we're not supposed to new up a DBContext like that and it's scoped per request. I'm not sure what per request means for a console app, and I think that's meant for .NET web API or web apps.

What am I doing wrong? How should I be passing the DBContext in? Should I be getting a new DBContext? Thanks so much in advance!

**Comment by djrpascu** [comment]
@ajcvickers Thanks so much for taking the time to reply. I would like to stick with using D.I. as this is more of an exercise for me to understand how to use it. How would I create a new scope for each pass through in the loop? Would I pass in the serviceProvider and do something like serviceProvider.CreateScope()? Do you have an example, or could you kindly point me to a sample console app that does the same? Much Mahalo!

**Comment by djrpascu** [comment]
@ajcvickers Did some digging, your suggestions lead to me use more helpful terms in my Google searches, so thanks!

I ended up injecting IServiceScopeFactory then creating scope like so:

```
                        using (var scope = _ServiceScopeFactory.CreateScope())
                        {
                            var db = scope.ServiceProvider.GetService<TOIDPContext>();

                            db.Entries.Add(entry);
                            db.SaveChanges();
                        }
```

That worked, but I see now that I really should be defining scope at service level rather than the DBContext.

So is my understanding correct in that for console apps, we need to manually define scope using IServiceScopeFactory, like above? I came across an article that noted web requests are wrapped in a scope, hence the lifetime observed in web apps/apis. Originally, I thought just calling a service method would give me new scope because of that, but that's just my lack of knowledge on how D.I. works.

Thanks for the help! If you have any tips or input, I'd appreciate it! I'll close this out. Mahalo!
